### PR TITLE
Xcode integration for nix

### PR DESCRIPTION
**Pull request by pikajude**
After several months of hacks, weird runtime bugs, and frustrations over files missing from Apple's open-source releases, I've become increasingly convinced that this is the most viable option for anyone who wishes to build Darwin-specific projects that extensively use system frameworks.

This fork adds constants `xcodeSDKRoot` and `xcodeVersion`, which are detected during evaluator initialization based on the current Xcode install.

In addition to this I have a fork https://github.com/pikajude/darwinixpkgs, which I don't think is ready to PR yet, but which features:
- A new darwin stdenv that integrates with Xcode
- A sandbox profile that allows access to a subset of SDK headers (since they include headers for Python, Ruby, MySQL, etc)
- First-class support for framework dependencies

### CODE REVIEW

**Comment by cstrahan** [comment]
Looks interesting!

Are you planning to put this into a state such that it can be merged (rather than maintained as a separate fork)?

How does this relate to the pure stdenv? Do you envision that the pure stdenv should be abandoned in favor of this?

**Comment by pikajude** [comment]
@cstrahan I think the pure stdenv should be an alternative but this should be the default.

I'll work on the merge conflict later tonight, but I wanted to get this PR opened to get some people's opinions first.

**Comment by veprbl** [comment]
@pikajude 
In that case you will have to have hydra builds for every supported OSX target, or you will have to throw the purity out of the window, I suppose. You also get occasional broken builds after SDK gets updated (like brew has).

**Comment by pikajude** [comment]
@veprbl Correct, I'd like to have a Hydra instance for each of the most recent 2 or 3 Xcode versions.

**Comment by veprbl** [comment]
@pikajude 
There also will be people wanting to use Command Line Tools instead of Xcode.

For completeness of this PR could you please name some applications you can get working this way and which closed-source frameworks they depend on?

**Comment by pikajude** [comment]
@veprbl xhyve (Kernel and hypervisor frameworks), anything touching NSString, emacs with proper Cocoa support.

The problem is not whether we can compile the programs but whether their runtime behavior will be predictable and correct. The headers that pure-darwin includes occasionally are inconsistent with the actual object files in ways that are very difficult to predict and debug—for example, compiling phantomjs2 with pure-darwin's CF headers would cause a runtime failure whenever NSString was used—and, more importantly, these inconsistencies vary between OS versions, meaning that we need *some* way of determining what libc version, at least, the end user is on.

I figured, since the official fix for phantomjs2 at the time was to relink it with `/System/Library/.../CoreFoundation` instead of our CoreFoundation, the line of purity was sufficiently blurry that I was willing to give it up a little more in exchange for a set of libraries and headers that are guaranteed to work together.

**Comment by pikajude** [comment]
@veprbl For the reason given above, I can't provide a fallback for people who want to use the CLT since it doesn't include the standard headers. They would need to use the pure-darwin stdenv instead.

**Comment by veprbl** [comment]
@pikajude 
This makes sense. But I don't understand the CLT problem. It can compile things so there are headers somewhere, maybe not presented as an SDK, but they are there, right?

**Comment by copumpkin** [comment]
To be clear, I think there's still some disagreement out there (at least with me) as to whether purity is worth throwing out just yet. I don't think we've really exhausted our options in that regard, and although I appreciate the contribution, I don't personally want darwinixpkgs to become the default. I don't have a strong opinion about this particular PR, since it's enabling darwinixpkgs as opposed to changing nixpkgs to switch stdenv.

Anyway, I acknowledge the difficulties you list as motivations, but as mentioned on IRC, I do think there are sensible ways to fix those while not throwing out the current approach. Getting people to do the work has been hard, but I don't think adding another stdenv for them to divide their attention with is going to fix that.

**Comment by edolstra** [comment]
Requiring Xcode seems like a distinct step back... I'm very happy that Nix on Darwin no longer requires Xcode and things (mostly) just work out of the box. I don't think our build machines even have Xcode installed at the moment.

**Comment by copumpkin** [comment]
Mostly agree with Charles here: I could definitely see the impure one being useful but I'm also afraid that it'll become infectious and people will start using it by default if they run into issues with the other one, which makes it harder for us to know about problems and improve the main one. Perhaps we could use the new lib.warn function to emit a little warning telling people  that they're using the impure stdenv mkDerivation and that there could be more system-dependent issues. Do you think that would be too annoying for end users?

**Comment by pikajude** [comment]
@edolstra Understood that this seems like a step back. Two things:

1. I intend the Xcode stdenv to be totally separate and opt-in, so the build machines wouldn't need Xcode installed in the first place—this could be seen as a "Darwin power users" feature.
2. Evaluation won't actually require xcode unless the value of one of the xcode* builtins is forced, so we could just check for their existence without causing a runtime error. (At least, I think that's the case. I'll check once I have some spare time to spend on this).

File: src/libexpr/primops.cc [hunk 1]
```diff
@@ -1793,6 +1793,37 @@ static void prim_fetchTarball(EvalState & state, const Pos & pos, Value * * args
     fetch(state, pos, args, v, "fetchTarball", true);
 }
 
+#if __APPLE__
+/*************************************************************
+ * Xcode integration
+ *************************************************************/
+
+std::string exec(const char* cmd) {
```
Comment (edolstra) [inline src/libexpr/primops.cc]: Please use `runProgram()` from util.cc which already does this.

File: src/libexpr/primops.cc [hunk 2]
```diff
@@ -1979,6 +2010,15 @@ void EvalState::createBaseEnv()
         for (auto & primOp : *RegisterPrimOp::primOps)
             addPrimOp(std::get<0>(primOp), std::get<1>(primOp), std::get<2>(primOp));
 
+#if __APPLE__
+    // Xcode integration
+    mkString(v, lookupXcodeVersion());
+    addConstant("__xcodeVersion", v);
+
+    mkString(v, lookupSDKRoot());
+    addConstant("__xcodeSDKRoot", v);
```
Comment (edolstra) [inline src/libexpr/primops.cc]: This is not lazy, so it will run two child processes at startup time, and throw an exception if Xcode is not installed. I think the easiest way to make this lazy is to turn it into a primop with one dummy argument, and then use `mkApp` to pass the dummy argument.

File: src/libstore/build.cc [hunk 3]
```diff
@@ -1680,7 +1680,7 @@ void DerivationGoal::startBuilder()
     {
         string x = settings.get("build-use-sandbox",
             /* deprecated alias */
-            settings.get("build-use-chroot", string("false")));
+            settings.get("build-use-chroot", string("relaxed")));
```
Comment (edolstra) [inline src/libstore/build.cc]: Changing the default sandbox mode probably belong in a separate PR.

File: src/libstore/build.cc [hunk 4]
```diff
@@ -1736,7 +1736,11 @@ void DerivationGoal::startBuilder()
 
     /* In a sandbox, for determinism, always use the same temporary
        directory. */
-    tmpDirInSandbox = useChroot ? canonPath("/tmp", true) + "/nix-build-" + drvName + "-0" : tmpDir;
+    tmpDirInSandbox =
+#if !__APPLE__
+      useChroot ? canonPath("/tmp", true) + "/nix-build-" + drvName + "-0" :
+#endif
```
Comment (edolstra) [inline src/libstore/build.cc]: Maybe add a comment explaining why this is conditional on `!__APPLE__`?

File: src/resolve-system-dependencies/resolve-system-dependencies.cc [hunk 5]
```diff
@@ -41,6 +43,11 @@ std::string findDylibName(bool should_swap, ptrdiff_t dylib_command_start) {
 }
 
 std::set<std::string> runResolver(const Path & filename) {
+    if(getFileType(filename) != DT_REG) {
+        printMsg(lvlError, format("%1%: not a regular file") % filename);
```
Comment (edolstra) [inline src/resolve-system-dependencies/resolve-system-dependencies.cc]: => `printError("%s: not a regular file", filename)`.

File: src/resolve-system-dependencies/resolve-system-dependencies.cc [hunk 6]
```diff
@@ -157,6 +185,10 @@ int main(int argc, char ** argv) {
     return handleExceptions(argv[0], [&]() {
         initNix();
 
+        if(argc < 2) {
+            throw Error(format("missing derivation argument"));
```
Comment (edolstra) [inline src/resolve-system-dependencies/resolve-system-dependencies.cc]: `format` is not needed here (or indeed almost anywhere).

File: release.nix [hunk 7]
```diff
@@ -27,7 +27,7 @@ let
           [ curl bison flex perl libxml2 libxslt bzip2 xz
             pkgconfig sqlite libsodium boehmgc
             docbook5 docbook5_xsl
-            autoconf-archive
+            autoconf-archive libyamlcpp
```
Comment (edolstra) [inline release.nix]: This should be conditional on `stdenv.isDarwin` since we don't need libyamlcpp elsewhere.

File: README.md [hunk 8]
```diff
@@ -1,14 +1,24 @@
...
+
 To find out more about the tool, usage and installation instructions, please
 read the manual, which is available on the Nix website at
 <http://nixos.org/nix/manual>.
 
+## Installing
+
+``` bash
+$ nix-env -f release.nix -iA build.x86_64-darwin
+```
```
Comment (edolstra) [inline README.md]: We shouldn't duplicate the manual here, so please don't add installation instructions here.

File: README.md [hunk 9]
```diff
@@ -1,14 +1,24 @@
...
 ------------------------------------------
 
-Nix is a new take on package management that is fairly unique. Because of it's
+Nix is a new take on package management that is fairly unique. Because of its
 purity aspects, a lot of issues found in traditional package managers don't
 appear with Nix.
 
+*Darwinix* is an OSX-specific port of nix that integrates more tightly with Xcode,
+avoiding a host of problems introduced by the fact that the vast majority of OSX's system
+libraries are closed-source.
```
Comment (edolstra) [inline README.md]: IMHO it's just confusing to introduce a separate name for Nix on Darwin.

File: corepkgs/buildenv.nix [hunk 10]
```diff
@@ -25,19 +25,16 @@ derivation {
   # Also don't bother substituting.
   allowSubstitutes = false;
 
+  __impureHostDeps = [
+    "/usr/lib/libSystem.dylib"
+  ];
```
Comment (edolstra) [inline corepkgs/buildenv.nix]: This should be conditional on Darwin.
